src/DataFramesMeta.jl

```diff
@@ -388,19 +387,86 @@ end
     for (k, v) in kwargs
         first = v(g[1])
-        result[k] = Array{eltype(first)}(undef, size(result, 1))
-        result[idx1[1]:idx2[1], k] = first
-        for i in 2:length(g)
-            result[idx1[i]:idx2[i], k] = v(g[i])
+        if first isa AbstractVector
+            t = _transform!(Tables.allocatecolumn(eltype(first), size(result, 1)), first, 1, g, v, starts, ends)
+        else 
+            t = _transform!(Tables.allocatecolumn(typeof(first), size(result, 1)), first, 1, g, v, starts, ends)
         end
+        result[k] = t
     end
     return result
 end
 
+function _transform!(t::AbstractVector, first::AbstractVector, start::Int, g::GroupedDataFrame, v::Function, starts::Vector, ends::Vector)
+    # handle the first case 
+    j = fill_column_vec!(t, first, starts[start], ends[start], size(g[start], 1))
+    @assert j === nothing
+    for i in (start+1):length(g)
+        out = v(g[i])
+        promoted = fill_column_vec!(t, out, starts[i], ends[i], size(g[i], 1))
+        if !(promoted === nothing)
+             t = copyto!(Tables.allocatecolumn(promoted, length(t)), 
+                         1, t, 1, ends[i-1])
+             _transform!(t, out, i, g, v, starts, ends)
+         end
+    end
+    return t
+end
+
+function _transform!(t::AbstractVector, first::Any, start::Int, g::GroupedDataFrame, v::Function, starts::Vector, ends::Vector)
+    # handle the first case 
+    j = fill_column_any!(t, first, starts[start], ends[start])
+    @assert j === nothing
+    for i in (start+1):length(g)
+        out = v(g[i])
+        promoted = fill_column_any!(t, out, starts[i], ends[i])
+        if !(promoted === nothing)
+             t = copyto!(Tables.allocatecolumn(promoted, length(t)), 
+                         1, t, 1, ends[i-1])
+             _transform!(t, out, i, g, v, starts, ends)
+         end
+    end
+    return t
+end
+
+function fill_column_vec!(t::AbstractVector, out, startpoint::Int, endpoint::Int, len::Int)
+    if !(out isa AbstractVector)
+         throw(ArgumentError("Return value must be an `AbstractVector` for all groups or for none of them"))
+    elseif length(out) != len
+        throw(ArgumentError("If a function returns a vector, the result " * 
+                            "must have the same length as the groups it operates on"))    
+    end     
+    elout = eltype(out)
+    T = eltype(t)
+    promoted = promote_type(elout, T)
+    if (elout <: T || promoted <: T)
+        t[startpoint:endpoint] = out
+    else 
+        return promoted
+    end
+    return nothing 
+end 
+
+function fill_column_any!(t::AbstractVector, out, startpoint::Int, endpoint::Int)
+    if (out isa AbstractVector)
+         throw(ArgumentError("Return value must be an `AbstractVector` for all groups or for none of them"))
+    end     
+    typout = typeof(out)
+    T = eltype(t)
+    promoted = promote_type(typout, T)
+    if (typout <: T || promoted <: T)
+        @views t[startpoint:endpoint] .= Ref(out)
```

Comment:
`@views` isn't needed on assignment.
```suggestion
        t[startpoint:endpoint] .= Ref(out)
```